meeteval/io/seglst.py

```diff
@@ -0,0 +1,388 @@
+import dataclasses
+import functools
+import typing
+from typing import List, Callable
+
+try:
+    from functools import cached_property
+except ImportError:
+    # Fallback for Python 3.7 and lower, since cached_property was added in
+    # Python 3.8.
+    from cached_property import cached_property  # Python 3.7
+
+from meeteval.io.base import BaseABC
+from meeteval.io.py import NestedStructure
+if typing.TYPE_CHECKING:
+    from meeteval.wer.wer.error_rate import ErrorRate
+
+
+try:
+    import TypedDict
+
+    class SegLstSegment(TypedDict, total=False):
+        """
+        A segment.
+
+        Note:
+            We do not define an enum with all these keys for speed reasons
+        """
+        session_id: str
+        start_time: float
+        end_time: float
+        words: str
+        speaker: str
+        segment_index: int
+
+        # Unused but by MeetEval but present in some file formats. They are defined here for compatibility and
+        # conversion in both directions
+        channel: int
+        confidence: float
+except ImportError:
+    # Python <3.8 doesn't have TypedDict
+    SegLstSegment = typing.Dict[str, typing.Any]
+
+
+@dataclasses.dataclass(frozen=True)
+class SegLST(BaseABC):
+    """
+    A collection of segments in SegLST format. This the input type to most
+    functions in MeetEval that process transcript segments.
+    """
+    segments: 'List[SegLstSegment]'
+
+    # Caches
+    _unique = None
+
+    @cached_property
+    def keys(self):
```

Comment:
How about something like the following?
```python
@property
class T:
    def __init__(self, outer):
        self._outer = outer
    def keys(self):
        return set.intersection(*[set(s.keys()) for s in self._outer.segments])
    def __getitem__(self, key):
        return [s[key] for s in self._outer.segments]
```
Do you have a suggestion for an alternative name? (I took `T` from numpy, where it means transpose)

```python
t = asseglst(['a b c', 'd e f', 'g h i'])
print(t.T.keys())  # 'words', 'segment_index', 'speaker'
print(t.T['words'])  # ['a b c', 'd e f', 'g h i']
```

Reply:
I like `t.T['words']` a lot. This operation is frequently used in the codebase but is currently always implemented manually with a list comprehension. From pandas we could use the name `columns`, but I think I like `T` better.

Reply:
I now realized, that keys is a property and not a function. With `cached_property` and caching inside the class, the second call could get a significant speedup. How likely is it, that a user manipulates a SegLST, while it should be treated as immutable?

Reply:
So the `cached_property` on this keys method may introduce bugs?

Ok, then, no caching sound better.

Maybe we should add a comment to that class, that it should be treated as immutable. I am not sure, whether we will hit a performance issue without caching. (In lazy_dataset I had to add code to cache the keys, because it was called too often. I guess here it will not be the case.)

Reply:
Okay, I removed the caching. We can keep an eye on this and add some sort of caching when we notice performance issues